File: sdk/ruby/test/wire/messages_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class MessagesWireTest < WireMockTestCase
  def setup
    super

    @client = AgentMailer::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_messages_list_with_wiremock
    test_id = "messages.list.0"

    @client.messages.list(
      inbox_id: "inboxId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "messages.list.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/inboxes/inboxId/messages",
      query_params: nil,
      expected: 1
    )
  end

  def test_messages_send__with_wiremock
    test_id = "messages.send_.0"

    @client.messages.send_(
      inbox_id: "inboxId",
      to: ["to"],
      request_options: {
        additional_headers: {
          "X-Test-Id" => "messages.send_.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/inboxes/inboxId/messages/send",
      query_params: nil,
      expected: 1
    )
  end

  def test_messages_get_with_wiremock
    test_id = "messages.get.0"

    @client.messages.get(
      inbox_id: "inboxId",
      message_id: "messageId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "messages.get.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/inboxes/inboxId/messages/messageId",
      query_params: nil,
      expected: 1
    )
  end

  def test_messages_delete_with_wiremock
    test_id = "messages.delete.0"

    @client.messages.delete(
      inbox_id: "inboxId",
      message_id: "messageId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "messages.delete.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "DELETE",
      url_path: "/v1/inboxes/inboxId/messages/messageId",
      query_params: nil,
      expected: 1
    )
  end

  def test_messages_reply_with_wiremock
    test_id = "messages.reply.0"

    @client.messages.reply(
      inbox_id: "inboxId",
      message_id: "messageId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "messages.reply.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/inboxes/inboxId/messages/messageId/reply",
      query_params: nil,
      expected: 1
    )
  end

  def test_messages_reply_all_with_wiremock
    test_id = "messages.reply_all.0"

    @client.messages.reply_all(
      inbox_id: "inboxId",
      message_id: "messageId",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "messages.reply_all.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/inboxes/inboxId/messages/messageId/reply-all",
      query_params: nil,
      expected: 1
    )
  end

  def test_messages_forward_with_wiremock
    test_id = "messages.forward.0"

    @client.messages.forward(
      inbox_id: "inboxId",
      message_id: "messageId",
      to: ["to"],
      request_options: {
        additional_headers: {
          "X-Test-Id" => "messages.forward.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/inboxes/inboxId/messages/messageId/forward",
      query_params: nil,
      expected: 1
    )
  end
end

File: sdk/rust/tests/messages_test.rs
use agentmailer::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_messages_list_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .messages
        .list(
            &"inboxId".to_string(),
            &MessagesListQueryRequest {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/inboxes/inboxId/messages", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_messages_send_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .messages
        .send(
            &"inboxId".to_string(),
            &MessageCompose {
                to: vec![MessageComposeToItem::String("to".to_string())],
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("POST", "/v1/inboxes/inboxId/messages/send", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_messages_get_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .messages
        .get(&"inboxId".to_string(), &"messageId".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count("GET", "/v1/inboxes/inboxId/messages/messageId", None, 1)
        .await
        .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_messages_delete_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .messages
        .delete(&"inboxId".to_string(), &"messageId".to_string(), None)
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "DELETE",
        "/v1/inboxes/inboxId/messages/messageId",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_messages_reply_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .messages
        .reply(
            &"inboxId".to_string(),
            &"messageId".to_string(),
            &ReplyCompose {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/inboxes/inboxId/messages/messageId/reply",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_messages_reply_all_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .messages
        .reply_all(
            &"inboxId".to_string(),
            &"messageId".to_string(),
            &ReplyCompose {
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/inboxes/inboxId/messages/messageId/reply-all",
        None,
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_messages_forward_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = AgentMailer::new(config).expect("Failed to build client");

    let result = client
        .messages
        .forward(
            &"inboxId".to_string(),
            &"messageId".to_string(),
            &MessageCompose {
                to: vec![MessageComposeToItem::String("to".to_string())],
                ..Default::default()
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/inboxes/inboxId/messages/messageId/forward",
        None,
        1,
    )
    .await
    .unwrap();
}
